[WebGPU] Unify Surface and SwapChain into a single class

litherum opened the pull request:
#### a5b1ddef1f3bfcf16d61b7c89f474ab1bc787395
<pre>
[WebGPU] Unify Surface and SwapChain into a single class
<a href="https://bugs.webkit.org/show_bug.cgi?id=250694">https://bugs.webkit.org/show_bug.cgi?id=250694</a>
rdar://104312969

Reviewed by Kimmo Kinnunen.

In ancient history, the WebGPU spec had a GPUCanvasContext as distinct from a GPUSwapChain.
However, since then, those two objects in the spec have been unified. This is a good thing;
you can&apos;t have multiple swap chains alive for the same canvas context, and the canvas context
itself didn&apos;t actually have any methods other than one to let you create a swap chain.

For those same reasons, and because splitting the context across 2 classes is difficult to
understand, we can follow the same model, and unify Surface and SwapChain into the same class.
The name of the new unified class is PresentationContext. When a client wants to create a swap
chain, all we do is bump the reference count of the PresentationContext.

No test because there is no behavior change.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(wgpuDeviceCreateSwapChain):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::createSurface):
(wgpuInstanceCreateSurface):
* Source/WebGPU/WebGPU/PresentationContext.h: Renamed from Source/WebGPU/WebGPU/Surface.h.
(WebGPU::PresentationContext::create):
(WebGPU::PresentationContext::displayBuffer const):
(WebGPU::PresentationContext::drawingBuffer const):
* Source/WebGPU/WebGPU/PresentationContext.mm: Renamed from Source/WebGPU/WebGPU/Surface.mm.
(WebGPU::optionsFor32BitSurface):
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::Device::createSurface):
(WebGPU::Device::createSwapChain):
(WebGPU::PresentationContext::PresentationContext):
(WebGPU::PresentationContext::getPreferredFormat):
(WebGPU::PresentationContext::configure):
(WebGPU::PresentationContext::present):
(WebGPU::PresentationContext::getCurrentTextureView):
(WebGPU::PresentationContext::nextDrawable):
(wgpuSurfaceRelease):
(wgpuSwapChainRelease):
(wgpuSurfaceGetPreferredFormat):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):
(wgpuSwapChainGetCurrentTextureView):
(wgpuSwapChainPresent):
* Source/WebGPU/WebGPU/SwapChain.h: Removed.
* Source/WebGPU/WebGPU/SwapChain.mm: Removed.

Canonical link: <a href="https://commits.webkit.org/258984@main">https://commits.webkit.org/258984@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3963ad35ec90c0f1a650329c35d382d316b6955

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112740 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3525 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111919 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109278 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38244 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92328 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79888 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6008 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26574 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6185 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3101 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12167 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46083 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6172 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7940 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/a3963ad35ec90c0f1a650329c35d382d316b6955)<details>

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103504 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12622 "Build is in progress. Recent messages:Running worker_preparation") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36465 "Build is in progress. Recent messages:Running worker_preparation") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112740 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95759 "Build is in progress. Recent messages:") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/36465 "Build is in progress. Recent messages:Running worker_preparation") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/95759 "Build is in progress. Recent messages:") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/36465 "Build is in progress. Recent messages:Running worker_preparation") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/95759 "Build is in progress. Recent messages:") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/36465 "Build is in progress. Recent messages:Running worker_preparation") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3750 "Build is in progress. Recent messages:Running configure-build") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6185 "Build is in progress. Recent messages:") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/102384 "Build is in progress. Recent messages:configuring build") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12167 "Build is in progress. Recent messages:Running validate-change") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/36465 "Build is in progress. Recent messages:Running worker_preparation") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/98673 "Build is in progress. Recent messages:Running configure-build") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7940 "Build is in progress. Recent messages:Running worker_preparation") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/37/builds/98673 "Build is in progress. Recent messages:Running configure-build") | 
| | | | | </details>

mwyrzykowski reviewed (approved):
This is really great, thank you for the refactor

litherum committed:
[WebGPU] Unify Surface and SwapChain into a single class
https://bugs.webkit.org/show_bug.cgi?id=250694
rdar://104312969

Reviewed by Kimmo Kinnunen.

In ancient history, the WebGPU spec had a GPUCanvasContext as distinct from a GPUSwapChain.
However, since then, those two objects in the spec have been unified. This is a good thing;
you can't have multiple swap chains alive for the same canvas context, and the canvas context
itself didn't actually have any methods other than one to let you create a swap chain.

For those same reasons, and because splitting the context across 2 classes is difficult to
understand, we can follow the same model, and unify Surface and SwapChain into the same class.
The name of the new unified class is PresentationContext. When a client wants to create a swap
chain, all we do is bump the reference count of the PresentationContext.

No test because there is no behavior change.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(wgpuDeviceCreateSwapChain):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::createSurface):
(wgpuInstanceCreateSurface):
* Source/WebGPU/WebGPU/PresentationContext.h: Renamed from Source/WebGPU/WebGPU/Surface.h.
(WebGPU::PresentationContext::create):
(WebGPU::PresentationContext::displayBuffer const):
(WebGPU::PresentationContext::drawingBuffer const):
* Source/WebGPU/WebGPU/PresentationContext.mm: Renamed from Source/WebGPU/WebGPU/Surface.mm.
(WebGPU::optionsFor32BitSurface):
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::Device::createSurface):
(WebGPU::Device::createSwapChain):
(WebGPU::PresentationContext::PresentationContext):
(WebGPU::PresentationContext::getPreferredFormat):
(WebGPU::PresentationContext::configure):
(WebGPU::PresentationContext::present):
(WebGPU::PresentationContext::getCurrentTextureView):
(WebGPU::PresentationContext::nextDrawable):
(wgpuSurfaceRelease):
(wgpuSwapChainRelease):
(wgpuSurfaceGetPreferredFormat):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):
(wgpuSwapChainGetCurrentTextureView):
(wgpuSwapChainPresent):
* Source/WebGPU/WebGPU/SwapChain.h: Removed.
* Source/WebGPU/WebGPU/SwapChain.mm: Removed.

Canonical link: https://commits.webkit.org/258984@main

webkit-commit-queue commented:
Committed 258984@main (a5b1ddef1f3b): <https://commits.webkit.org/258984@main>

Reviewed commits have been landed. Closing PR #8708 and removing active labels.
